src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+{
+}
+
+ICeWall::~ICeWall()
+{
+}
+
+void ICeWall::init(odb::dbDatabase* db, utl::Logger* logger)
+{
+  db_ = db;
+  logger_ = logger;
+}
+
+odb::dbBlock* ICeWall::getBlock() const
+{
+  auto* chip = db_->getChip();
+  if (chip == nullptr) {
+    return nullptr;
+  }
+
+  return chip->getBlock();
+}
+
+void ICeWall::makeBumpArray(odb::dbMaster* master,
+                            const odb::Point& start,
+                            int rows,
+                            int columns,
+                            int xpitch,
+                            int ypitch,
+                            const std::string& prefix)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (master == nullptr) {
+    return;
+  }
+
+  for (int xn = 0; xn < columns; xn++) {
+    for (int yn = 0; yn < rows; yn++) {
+      const odb::Point pos(start.x() + xn * xpitch, start.y() + yn * ypitch);
+      const std::string name = fmt::format("{}{}_{}", prefix, xn, yn);
+      auto* inst = odb::dbInst::create(block, master, name.c_str());
+
+      inst->setOrigin(pos.x(), pos.y());
+      inst->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+    }
+  }
+}
+
+void ICeWall::removeBump(odb::dbInst* inst)
+{
+  if (inst == nullptr) {
+    return;
+  }
+
+  odb::dbInst::destroy(inst);
+}
+
+void ICeWall::removeBumpArray(odb::dbMaster* master)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (master == nullptr) {
+    return;
+  }
+
+  for (auto* inst : block->getInsts()) {
+    if (inst->getMaster() == master) {
+      removeBump(inst);
+    }
+  }
+}
+
+void ICeWall::makeBTerm(odb::dbNet* net,
+                        odb::dbTechLayer* layer,
+                        const odb::Rect& shape) const
+{
+  odb::dbBTerm* bterm = odb::dbBTerm::create(net, net->getConstName());
+  if (bterm == nullptr) {
+    bterm = net->get1stBTerm();
+  }
+  bterm->setSigType(net->getSigType());
+  odb::dbBPin* pin = odb::dbBPin::create(bterm);
+  odb::dbBox::create(
+      pin, layer, shape.xMin(), shape.yMin(), shape.xMax(), shape.yMax());
+  pin->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+}
+
+void ICeWall::assignBump(odb::dbInst* inst, odb::dbNet* net)
+{
+  if (inst == nullptr) {
+    return;
+  }
+  if (net == nullptr) {
+    return;
+  }
+
+  odb::dbTransform xform;
+  inst->getTransform(xform);
+
+  // Connect to all iterms since this is a bump
+  for (auto* pin : inst->getITerms()) {
+    if (pin->getNet() != net) {
+      pin->connect(net);
+    }
+
+    for (auto* mpin : pin->getMTerm()->getMPins()) {
+      for (auto* geom : mpin->getGeometry()) {
+        auto* layer = geom->getTechLayer();
+        if (layer == nullptr) {
+          continue;
+        }
+
+        odb::Rect shape = geom->getBox();
+        xform.apply(shape);
+        makeBTerm(net, layer, shape);
+      }
+    }
+  }
+}
+
+void ICeWall::makeFakeSite(const std::string& name, int width, int height)
```

Comment:
What about symmetry settings?

Reply:
For the fake sites it's really just there to be able to make the rows, I could add the symmetry if needed.